include/measurement_kit/nettest.hpp

```diff
@@ -0,0 +1,2589 @@
+    virtual void on_status_geoip_lookup(events::StatusGeoipLookupEvent);
+
+    /// Handles the "status.progress" event.
+    virtual void on_status_progress(events::StatusProgressEvent);
+
+    /// Handles the "status.queued" event.
+    virtual void on_status_queued(events::StatusQueuedEvent);
+
+    /// Handles the "status.measurement_start" event.
+    virtual void on_status_measurement_start(events::StatusMeasurementStartEvent);
+
+    /// Handles the "status.measurement_submission" event.
+    virtual void on_status_measurement_submission(events::StatusMeasurementSubmissionEvent);
+
+    /// Handles the "status.measurement_done" event.
+    virtual void on_status_measurement_done(events::StatusMeasurementDoneEvent);
+
+    /// Handles the "status.report_close" event.
+    virtual void on_status_report_close(events::StatusReportCloseEvent);
+
+    /// Handles the "status.report_create" event.
+    virtual void on_status_report_create(events::StatusReportCreateEvent);
+
+    /// Handles the "status.resolver_lookup" event.
+    virtual void on_status_resolver_lookup(events::StatusResolverLookupEvent);
+
+    /// Handles the "status.started" event.
+    virtual void on_status_started(events::StatusStartedEvent);
+
+    /// Handles the "task_terminated" event.
+    virtual void on_task_terminated(events::TaskTerminatedEvent);
+
+  protected:
+    // Start a nettest given JSON settings
+    bool run_with_json_settings(nlohmann::json doc) noexcept;
+
+    // Dispatch the JSON event to the proper handler
+    virtual bool dispatch_event(nlohmann::json doc) noexcept;
+
+  private:
+    std::mutex mutex_;
+    detail::UniqueTask task_;
+};
+
+/// Base class for nettests measuring performance.
+class PerformanceNettest : public Nettest {
+  public:
+    /// Wait for nettest to terminate and destroy resources.
+    ~PerformanceNettest() noexcept override {}
+
+    /// Handles the "status.update.performance" event.
+    virtual void on_status_update_performance(events::StatusUpdatePerformanceEvent);
+
+  protected:
+    // Dispatch the JSON event to the proper handler
+    bool dispatch_event(nlohmann::json doc) noexcept override;
+};
+
+/// Base class for nettest measuring websites blocking.
+class WebsitesNettest : public Nettest {
+  public:
+    /// Wait for nettest to terminate and destroy resources.
+    ~WebsitesNettest() noexcept override {}
+
+    /// Handles the "status.update.websites" event.
+    virtual void on_status_update_websites(events::StatusUpdateWebsitesEvent);
+
+  protected:
+    // Dispatch the JSON event to the proper handler
+    bool dispatch_event(nlohmann::json doc) noexcept override;
+};
+
+} // namespace common
+
+#ifndef MK_NETTEST_NO_CAPTIVE_PORTAL
+
+/// Settings for CaptivePortal
+class CaptivePortalSettings : public common::Settings {
+  public:
+    /// Unique name of this nettest.
+    static constexpr const char *name = "CaptivePortal";
+    /* No nettest-specific settings */
```

Comment:
I have not mapped settings for most nettests yet. In most cases there are nettest specific settings.